### ISSUE TITLE
Text input / IME support

### ISSUE THREAD

**Issue by tehsausage** [liballeg/allegro5#1191]
Basing this idea off of SDL's implementation: https://wiki.libsdl.org/Tutorials/TextInput

### Problem

There is no way to input "international" characters in to Allegro programs using an [IME](https://en.wikipedia.org/wiki/Input_method).

This would be a desirable feature to have for games with any chat or text input features, or when developing tools for games where text may need to be entered.

### Proposed Feature

Add an API to enable "text input mode" for a display.

In this mode, raw keyboard input no longer necessarily maps directly to an Allegro keyboard event, and the function of some keys may be hijacked temporarily by the IME for inputting "compositions".

For example: Typing most letters while using [fcitx](https://fcitx-im.org/) will begin a composition. During composition, the directional arrow keys, home/end, space, enter, tab and escape keys, as well as other keys such as F7, are all hijacked for composition editing purposes, and not delivered to the program when pressed.

The intended use would be to enable "text input mode" while keyboard focus in an Allegro program is logically inside of a text input field. The program should respond to the receipt of `ALLEGRO_EVENT_TEXT_COMPOSITION` events by updating the display of their text input field as appropriate to show the composition-in-progress (conventionally: by displaying the composition's characters with an underline in the location where text is to be inserted), before finally receiving an `ALLEGRO_EVENT_TEXT_INPUT` event for the final output.

A program should consider that a composition can occur while the cursor is positioned in the middle of existing text. The composition could be displayed simply overlapping existing text (though care should be taken to wrap around at the edge of the window), or inserted in such a way that it affects word wrapping and layout as it changes. Clicking to move the cursor may either cancel the composition (call `al_stop_text_input` followed by `al_start_text_input`), or re-position the composition within the text. Care should be taken to not count the composition's characters towards the character count of the input field, and it should behave consistently when using the mouse to re-position the cursor.

If a particular IME or platform cannot provide composition events, it may operate silently without sending `ALLEGRO_EVENT_TEXT_COMPOSITION` events, and only emitting the final `ALLEGRO_EVENT_TEXT_INPUT` event.

Note that in practice it is not uncommon to see games which do not display compositions, thus handling of `ALLEGRO_EVENT_TEXT_COMPOSITION` events to display compositions is optional, even if the feature is used.

#### `al_start_text_input(ALLEGRO_DISPLAY* dpy)`

Enables "text input mode" on the display. Some keyboard input events may now be captured by the the IME.

#### `al_stop_text_input(ALLEGRO_DISPLAY* dpy)`

Disables "text input mode" on the display. 

Unclear what calling this should do if there is a composition in progress. It needs to be determined if all IME implementations can be instructed to "cancel" the current composition.

_(It might be more conventional to name this pair of functions `al_set_text_input_mode(ALLEGRO_DISPLAY*, bool)`)_

#### Event: `ALLEGRO_EVENT_TEXT_COMPOSITION`

**Parameters:**

* `display`: The display that text is being input on
* `text`: An `ALLEGRO_USTR` of the current text composition. If length is 0, then composition is effectively cancelled.

This is received for a display when text input composition begins or is updated.

**Big Issue:** Its not clear how "text" should be allocated and cleaned up. There's currently no built-in Allegro events which require explicit clean-up. Text input handling is happening asynchronously to event handling, so the text must necessarily be dynamically allocated.

A transparent solution could be to have `al_get_event()` call a destructor for the previous event, but that could lead to surprising lifetime errors.

#### Event: `ALLEGRO_EVENT_TEXT_INPUT`

**Parameters:**

* `display`: The display that text is being input on
* `text`: An `ALLEGRO_USTR` of the text that was finally entered

This is received for a display when text input composition completes, and the final characters are committed.

This event could be omitted, and instead a sequence of `ALLEGRO_KEY_CHAR` events could be used in its place. This would eliminate the concern about the lifetime of the string for this event, at least, but risks overflowing the event queue, and causing partial strings to be written.

#### `al_set_text_input_area(int x, int y, int w, int h)` (maybe)

Defines the size and location of the text input field, so that the platform knows where it can draw IME UI pop-ups, such as is supported on Windows:

![image](https://user-images.githubusercontent.com/8984972/99033527-048efa80-25cb-11eb-85e8-a3e4f723059c.png)

Will only take effect while in "text input mode". Setting a width and height of 0 would disable the effects of the function (I would expect there be a way to instruct platforms to not display UI elements at all).

A multi-line text input field should use the height of 1 line, or perhaps use the bounding box of the composition text display. Its not clear if updating this necessarily causes any currently displayed UI element to move.

### Platforms:

* **Windows**: Win32 has some support for transparently handling its IME, so a partial implementation could be to disable DirectInput keyboard handling in favour of WM_CHAR window events for as long as "text input mode" is active. Need to investigate how full IME support could be integrated. Windows 10 also has a win+period shortcut to open an emoji selector that should work with the same interface.

* **Linux**: As X11 is the only back-end, XIM would probably need to be used? I'm not certain how well XIM works, or if most common IME setups even configure XIM support. Ideally there would be toolkit-specific backend support added to Allegro so that GTK and Qt accessibility APIs could be used, as they seem to be much better supported.

* **Android and iOS**: This API could possibly be used to trigger the touch-screen keyboard to show up, as there doesn't seem to be a way to do this in Allegro at the moment? `ALLEGRO_EVENT_TEXT_INPUT` would also be convenient when entering complete words at a time. I'm not interested in working on mobile platforms, however.

* **Others**: I haven't considered any other platforms. Platforms that use the SDL2 backend should easily be supported through the SDL TextInput API, however.

---

### Example

Example of the events a program might receive while in text input mode, and how it might choose to display the composition in progress:

![kanji](https://user-images.githubusercontent.com/8984972/99036162-0bb90700-25d1-11eb-9dbc-38e46ec1cd4b.gif)

Step 1:
* User presses 'k'
* Allegro event: `ALLEGRO_EVENT_TEXT_COMPOSITION(text="ｋ", cursor=1)`

Step 2:
* User presses 'a'
* Allegro event: `ALLEGRO_EVENT_TEXT_COMPOSITION(text="か", cursor=1)`
* (note: cursor is still 1 here, based on the number of characters, not bytes, in the composition)

Step 3:
* User presses 'n'
* Allegro event: `ALLEGRO_EVENT_TEXT_COMPOSITION(text="かｎ", cursor=2)`

Step 4:
* User presses 'j'
* Allegro event: `ALLEGRO_EVENT_TEXT_COMPOSITION(text="かんｊ", cursor=3)`

Step 5:
* User presses 'i'
* Allegro event: `ALLEGRO_EVENT_TEXT_COMPOSITION(text="かんじ", cursor=3)`

Step 6:
* User presses left arrow
* Allegro event: `ALLEGRO_EVENT_TEXT_COMPOSITION(text="かんじ", cursor=2)`
* (note: the cursor here has moved within the composition, the program should update their display accordingly if appropriate)

Step 7:
* User presses space
* Allegro event: `ALLEGRO_EVENT_TEXT_COMPOSITION(text="感じ", cursor=0)`
* (note: the cursor has moved to the beginning of the string -- this is a quirk of fcitx)

Step 8:
* User presses space
* Allegro event: `ALLEGRO_EVENT_TEXT_COMPOSITION(text="漢字", cursor=0)`

Step 9:
* User presses enter
* Allegro event: `ALLEGRO_EVENT_TEXT_INPUT(text="漢字")`
* (note: the program may want to treat this as two ALLEGRO_KEY_CHAR events in sequence with unichar='漢' and unichar='字' respectively)
* (note: The receipt of an ALLEGRO_EVENT_TEXT_INPUT event means that the composition is complete, and should no longer be rendered by the program)

**Comment by beoran** [comment]
Yes, this is a great idea, and I looked into it before, but I found this is not very easy to implement, at least not on Linux, especially because XIM is very old and gnarly. As for the memory freeing problem, an al_acknowledge_text_event() function could be used to free the buffer and send acknowledgement to the IME. There is a precedent in Allegro for acknowledging events.

**Comment by divinciest** [comment]
I suggest the following : 
- add ALLEGRO_IME  struct
- add ALLEGRO_IME_EVENT struct
- add ALELGRO_IME_EVENT_SOURCE struct
- al_add_event_source(queue,  al_get_ime_event_source(ime)))

and in the event switch statement the user can handle the events ...

**Comment by LunarWatcher** [comment]
There's even more complexity on Linux unfortunately. XIM isn't the only IME on Linux. There's also at least IBus and fcitx5, plus several extra less-used ones. For what it's worth, the [Arch wiki](https://wiki.archlinux.org/title/Input_method) states

> The most common IMF is [IBus](https://wiki.archlinux.org/title/IBus) (often used in [GTK](https://wiki.archlinux.org/title/GTK)-based environments like [GNOME](https://wiki.archlinux.org/title/GNOME)), followed by [Fcitx5](https://wiki.archlinux.org/title/Fcitx5) (mostly used in [Qt](https://wiki.archlinux.org/title/Qt)-based environments like [KDE](https://wiki.archlinux.org/title/KDE)), [...]

So it'd probably be more impactful to go after IBus and/or Fcitx5 than XIM. I'm not sure how IBus and Fcitx5 are to work with compared to XIM though (never tried any of them).

**Comment by CoelacanthusHex** [comment]
> There's even more complexity on Linux unfortunately. XIM isn't the only IME on Linux. There's also at least IBus and fcitx5, plus several extra less-used ones.

Implement XIM and text-input-v3 is enough for most case. Both IBus and Fcitx supports these two protocols. The private DBus protocols of IBus and Fcitx are better to have but not necessary on x11. XIM will always work on X11 and text-input-v3 will always work on Wayland. On wayland, the private protocols are even worse than text-input-v3.
